### PR TITLE
UniqueConstraint: add test cases

### PR DESCRIPTION
**Pull request by swhite2**
Since UniqueConstraint was worked on recently, some test cases to accomodate all scenarios

### CODE REVIEW

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        list($ret, $msgs) = $container->validate();
+
+        $this->assertEquals(0, $msgs->count());
+        $this->assertEquals(true, $ret);
+    }
+
+    public function testEmptyValuesNonRequired()
+    {
+        $container = new UniqueTestContainer();
+        $container->setRequired(false);
```
Comment (fichtner) [inline src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:145]: given Required=N is the default this was hard to understand at first

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        $container->addNode(['unique_test' => 'value1']);
+        $container->addNode(['unique_test' => 'value1']);
+
+        list($ret, $msgs) = $container->validate();
+
+        $this->assertEquals(1, $msgs->count());
+        $this->assertEquals(false, $ret);
+    }
+
+    public function testNonEqualValues()
```
[the remark this answers is not in the document]
Comment (swhite2) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php]: @fichtner how about https://github.com/opnsense/core/pull/6875/commits/5ab6441a5a32cb08bf7f41b92d3a908b94e58dcb?

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        $container->addNode(['unique_test' => '']);
+
+        $msgs = $container->validate();
+
+        $this->assertEquals(0, $msgs->count());
+
+        $container = new UniqueTestContainer();
+        $container->setRequired(true);
+        $container->addNode(['unique_test' => '']);
+        $container->addNode(['unique_test' => '']);
```
Comment (fichtner) [inline src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:138]:
```suggestion
        $container->setRequired(true);
```

shouldn't this work? :)

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        $container->addNode(['unique_test' => '', 'unique_test2' => '']);
+
+        $msgs = $container->validate();
+
+        $this->assertEquals(0, $msgs->count());
+
+        $container = new UniqueTestContainer();
+        $container->setRequired(true);
+        $container->addNode(['unique_test' => '', 'unique_test2' => '']);
+        $container->addNode(['unique_test' => '', 'unique_test2' => '']);
```
Comment (fichtner) [inline src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:158]:
```suggestion
        $container->setRequired(true);
```

Same for this?

**Review by fichtner — APPROVED** [review]
look good minus minor comments

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php (hunk 3 above)
Comment (swhite2) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:138]: if we change the logic of UniqueTestContainer, yes

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+    public function testFirstValueEmptyPassAll()
+    {
+        $container = new UniqueTestContainer();
+        $container->addNode(['unique_test1' => '', 'unique_test2' => 'value1']);
+        $container->addNode(['unique_test1' => '', 'unique_test2' => 'value1']);
+
+        $msgs = $container->validate();
+
+        $this->assertEquals(0, $msgs->count());
```
[the remark this answers is not in the document]
Comment (swhite2) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:173]: You could argue that this is already covered by the earlier test, but I suppose with `addFields` behavior might be different

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php (hunk 3 above)
Comment (fichtner) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:138]: only if it's no trouble, don't need to spend more time than necessary

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php (hunk 5 above)
[the remark this answers is not in the document]
Comment (fichtner) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php:173]: might not be at the moment but better have an extra test that fails if the implementation is changed subtly than not catching it.

File: src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php (hunk 3 above)
Comment (swhite2) [reply src/opnsense/mvc/tests/app/models/OPNsense/Base/Constraints/UniqueConstraintTest.php]: https://github.com/opnsense/core/pull/6875/commits/4e05bcfc7506b24ca2489e14facf46ca361fd174 ;)

**Pull request merged**